src/sync.rs

```diff
@@ -100,25 +110,64 @@ unsafe impl<T> Sync for GILProtected<T> where T: Send {}
 /// }
 /// # Python::with_gil(|py| assert_eq!(get_shared_list(py).len(), 0));
 /// ```
-#[derive(Default)]
-pub struct GILOnceCell<T>(UnsafeCell<Option<T>>);
+pub struct GILOnceCell<T> {
+    once: Once,
+    data: UnsafeCell<MaybeUninit<T>>,
+
+    /// (Copied from std::sync::OnceLock)
+    ///
+    /// `PhantomData` to make sure dropck understands we're dropping T in our Drop impl.
+    ///
+    /// ```
+    /// use pyo3::Python;
+    /// use pyo3::sync::GILOnceCell;
+    ///
+    /// struct A<'a>(#[allow(dead_code)] &'a str);
+    ///
+    /// impl<'a> Drop for A<'a> {
+    ///     fn drop(&mut self) {}
+    /// }
+    ///
+    /// let cell = GILOnceCell::new();
+    /// {
+    ///     let s = String::new();
+    ///     let _ = Python::with_gil(|py| cell.set(py,A(&s)));
```

Comment:
One of the builds is failing here:

```
error[E0597]: `s` does not live long enough
  --> src/sync.rs:142:50
   |
24 |     let s = String::new();
   |         - binding `s` declared here
25 |     let _ = Python::with_gil(|py| cell.set(py,A(&s)));
   |                              ----                ^ borrowed value does not live long enough
   |                              |
   |                              value captured here
26 | }
   | - `s` dropped here while still borrowed
27 | } _doctest_main_src_sync_rs_121_0() }
   | - borrow might be used here, when `cell` is dropped and runs the `Drop` code for type `GILOnceCell`
```

Reply:
Ah my mistake, this is meant to be marked `compile_fail` as a UI test that the type behaves properly (imported from `std`).